tt_metal/tt_metal.cpp

```diff
@@ -554,6 +557,49 @@ void CloseDevices(std::map<chip_id_t, Device *> devices) {
         ZoneScoped;
         program.compile(device);
     }
+
+    void AllocateBuffer(Buffer* buffer, bool bottom_up) {
+        uint32_t allocated_addr;
+        if (buffer->device()->sw_command_queues_.size() == 0) {
+            // Using Slow dispatch since SW Command Queues are not initialized on device
+            detail::DispatchStateCheck(false);
+            if (is_sharded(buffer->buffer_layout())) {
+                allocated_addr = allocator::allocate_buffer(*(buffer->device()->allocator_), buffer->size(), buffer->page_size(), buffer->buffer_type(), bottom_up, buffer->num_cores());
+            }
+            else {
+                allocated_addr = allocator::allocate_buffer(*(buffer->device()->allocator_), buffer->size(), buffer->page_size(), buffer->buffer_type(), bottom_up, std::nullopt);
+            }
+            buffer->set_address(static_cast<uint64_t>(allocated_addr));
+        }
+        else {
+            // Using Fast dispatch. Push Allocate Command to CQ
+            detail::DispatchStateCheck(false);
+            EnqueueAllocateBuffer(buffer->device()->command_queue(), buffer, bottom_up, false);
+        }
+    }
+
+    void DeallocateBuffer(Buffer *buffer) {
+        if (buffer->device()->sw_command_queues_.size() == 0) {
+            // Using Slow dispatch since SW Command Queues are not initialized on device
+            detail::DispatchStateCheck(false);
+            allocator::deallocate_buffer(*(buffer->device()->allocator_), buffer->address(), buffer->buffer_type());
+        }
+        else {
+            // Using Fast dispatch. Push Deallocate Command to CQ
+            detail::DispatchStateCheck(false);
+            EnqueueDeallocateBuffer(buffer->device()->command_queue(), *(buffer->device()->allocator_), buffer->address(), buffer->buffer_type(), false);
+        }
+    }
+
+    void GetBufferAddress(const Buffer* buffer, uint32_t* address_on_host) {
+        if (buffer->device()->sw_command_queues_.size() == 0) {
+            *address_on_host = buffer->address();
+        }
+        else {
+            EnqueueGetBufferAddr(buffer->device()->command_queue(), address_on_host, buffer, false);
+        }
+    }
+
```

Comment:
same suggestion as the AssignGlobalBufferToProgram comment to avoid duplication of code here and in *Impl